posix/unix.c

```diff
@@ -815,11 +827,54 @@ int unix_shutdown(unsigned socket, int how)
 }
 
 
-int unix_setsockopt(unsigned socket, int level, int optname, const void *optval, socklen_t optlen)
+/* TODO: support buffer reallocation */
+static int set_buffer_size(unixsock_t *s, int sz)
 {
+	if (sz < US_MIN_BUFFER_SIZE || sz > US_MAX_BUFFER_SIZE)
+		return -EINVAL;
+
+	if (s->buffer.data != NULL)
+		return -EINVAL;
+
+	s->buffer_sz = sz;
+
 	return 0;
 }
 
+int unix_setsockopt(unsigned socket, int level, int optname, const void *optval, socklen_t optlen)
+{
+	unixsock_t *s;
+	int err;
+
+	if ((s = unixsock_get(socket)) == NULL)
+		return -ENOTSOCK;
+
+	do {
+		if (level != SOL_SOCKET) {
+			err = -EINVAL;
+			break;
```

Comment:
Move this before `unixsock_get`, drop `do ... while (0)`

Reply:
I don't like idea of returning EINVAL for non-existing socket.